src/SignalR/clients/ts/signalr/src/HttpConnection.ts

```diff
@@ -501,3 +502,70 @@ export class HttpConnection implements IConnection {
 function transportMatches(requestedTransport: HttpTransportType | undefined, actualTransport: HttpTransportType) {
     return !requestedTransport || ((actualTransport & requestedTransport) !== 0);
 }
+
+export class TransportSendQueue {
+    private sending: boolean = false;
+    private buffer: any[] = [];
+    private promiseQueue?: Promise<void>;
+
+    public async send(transport: ITransport, data: string | ArrayBuffer): Promise<void> {
+        return this.sending ?
+            this.bufferData(transport, data) :
+            this.sendCore(transport, data);
+    }
+
+    private async sendCore(transport: ITransport, data: string | ArrayBuffer) {
+        this.sending = true;
+
+        let resolve: () => void;
+        this.promiseQueue = new Promise((r) => resolve = r);
+
+        try {
+            await transport.send(data);
+        } finally {
+            this.sending = false;
+            this.promiseQueue = undefined;
```

Comment:
We should set up a new promise in `send` if we aren't sending. Nothing really would be using this once this is no longer buffering.

Reply:
I agree. I don't know why you'd clear this.

An a tangentially related note, it might be nicer if it wasn't nullable, and it was initialized to a completed promise.

Reply:
Are `.then` callbacks run inline when a promise finishes?

If no, then there is a race here when a callback has been registered and another send comes in right after a send completes.